alns/select/SelectionScheme.py

```diff
@@ -1,101 +1,111 @@
 from abc import ABC, abstractmethod
-from typing import List, Tuple
+from typing import Optional, Tuple
 
 import numpy as np
 from numpy.random import RandomState
 
+from alns.State import State
+
 
 class SelectionScheme(ABC):
     """
-    Base class from which to implement a selection scheme.
+    Base class from which to implement an operator selection scheme.
 
     Parameters
     ----------
-    scores
-        A list of four non-negative elements, representing the weight
-        updates when the candidate solution results in a new global best
-        (idx 0), is better than the current solution (idx 1), the solution
-        is accepted (idx 2), or rejected (idx 3).
     num_destroy
         Number of destroy operators.
     num_repair
         Number of repair operators.
+    op_coupling
+        Optional keyword argument. Matrix that indicates coupling between
+        destroy and repair operators. Entry (i, j) is 1 if destroy operator i
+        can be used in conjunction with repair operator j and 0 otherwise.
     """
 
-    def __init__(self, scores: List[float], num_destroy: int, num_repair: int):
-        self._validate_arguments(scores, num_destroy, num_repair)
+    def __init__(
+        self,
+        num_destroy: int,
+        num_repair: int,
+        *,
+        op_coupling: Optional[np.ndarray] = None,
```

Comment:
I don't like `op_coupling` as part of the base selection scheme. I feel that the use case is too niche. For most optimization problems there's no specific reason to use coupled operators.

I think that we should not include this feature in the base selection scheme. To not waste the effort put into the operator coupling feature, we could make a simple `CoupledRandomSelect` scheme.

Reply:
It's useful for problems with two levels of decision making (e.g. build facilities -> route). I know at least one master student that's looking into something like this, and he uses this kind of coupling for that.

If we can support it without too much effort that's a big plus IMHO. If it's hard to do we should at least support the roulette wheel mechanism, since that's more or less the default for ALNS these days. Other selection schemes are then free to do something else.

Reply:
It can be supported with very little effort. Users simply have to pass-in the op_coupling matrix, which we no longer need to construct in ALNS. Let's keep this feature then.